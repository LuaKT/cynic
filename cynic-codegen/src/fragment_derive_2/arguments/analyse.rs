use super::parsing;

pub struct Output;

pub fn analyze(literals: Vec<parsing::FieldArgument>, schema: ()) -> Output {
    // Somehow implement this.
    // Though it's not that simple, as I really need more info than the schema parsing in this module
    // currently gives me.

    todo!("analyze not implemented")
}
