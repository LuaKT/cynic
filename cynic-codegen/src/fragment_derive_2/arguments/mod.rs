mod analyse;
mod parsing;

pub use parsing::{arguments_from_field_attrs, FieldArgument};
