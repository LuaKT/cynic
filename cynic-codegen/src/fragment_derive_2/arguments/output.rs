use proc_macro2::Span;
use quote::{quote, ToTokens, TokenStreamExt};
use syn::Token;

use super::analyse::{AnalysedArguments, ArgumentValue};

pub struct Output<'a> {
    pub(super) analysed: AnalysedArguments<'a>,
    pub(super) schema_module: syn::Path,
}

impl ToTokens for Output<'_> {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        if self.analysed.arguments.is_empty() {
            return;
        }

        let argument_module = self
            .analysed
            .schema_field
            .argument_module()
            .to_path(&self.schema_module);

        for arg in &self.analysed.arguments {
            let arg_marker = proc_macro2::Ident::from(arg.schema_field.marker_ident());
            let value = &arg.value;

            tokens.append_all(quote! {
                field_builder.argument::<#argument_module::#arg_marker>()
                #value;
            });
        }
    }
}

impl ToTokens for ArgumentValue<'_> {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        match self {
            ArgumentValue::Object(_) => todo!("obj"),
            ArgumentValue::List(_) => todo!("list"),
            ArgumentValue::Literal(lit) => tokens.append_all(quote! {
                .literal(#lit)
            }),
            ArgumentValue::Bool(b) => {
                let lit = syn::LitBool::new(*b, Span::call_site());
                tokens.append_all(quote! {
                    .literal(#lit)
                });
            }
            ArgumentValue::Variable(_) => todo!("variable"),
            ArgumentValue::Some(_) => todo!("some"),
            ArgumentValue::Null => todo!("null"),
        }
    }
}
