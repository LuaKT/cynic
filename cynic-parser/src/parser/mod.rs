use logos::{Logos, Span};
use rowan::{GreenNode, GreenNodeBuilder};

use crate::{
    lexer::Token,
    syntax::{
        SyntaxKind::{self, *},
        SyntaxNode,
    },
};

mod arguments;
mod fragments;

use arguments::{arguments, constant_value};
use fragments::{fragment, fragment_spread, inline_fragment};

/// The parse results are stored as a "green tree".
/// We'll discuss working with the results later
pub struct Parse {
    green_node: GreenNode,
    #[allow(unused)]
    errors: Vec<(String, Option<Span>)>,
}

pub fn parse(text: &str) -> Parse {
    let mut lexer = Token::lexer(text);
    let mut tokens = Vec::new();
    while let Some(token) = lexer.next() {
        tokens.push((token, lexer.slice(), lexer.span()));
    }
    tokens.reverse();

    Parser {
        tokens,
        builder: GreenNodeBuilder::new(),
        errors: Vec::new(),
    }
    .parse()
}

impl Parse {
    fn syntax(&self) -> SyntaxNode {
        SyntaxNode::new_root(self.green_node.clone())
    }
}

struct Parser<'source> {
    tokens: Vec<(Token, &'source str, Span)>,
    builder: GreenNodeBuilder<'static>,

    // TODO: Do I want spanned errors?
    errors: Vec<(String, Option<Span>)>,
}

impl<'source> Parser<'source> {
    fn parse(mut self) -> Parse {
        self.builder.start_node(ROOT.into());
        let mut current_error = false;
        loop {
            // Ok, so parsing out definitions here.
            match executable_def(&mut self) {
                Res::Eof => {
                    break;
                }
                Res::ExpectedExecutableDef => {
                    self.builder.start_node(ERROR.into());
                    if !current_error {
                        self.error("expected executable definition");
                        current_error = true;
                    }
                    self.bump();
                    // TODO: better nodes somehow
                    self.builder.finish_node();
                }
                Res::Ok => {
                    current_error = false;
                }
            }
        }
        self.skip_ws();
        self.builder.finish_node();

        Parse {
            green_node: self.builder.finish(),
            errors: self.errors,
        }
    }

    fn current(&self) -> Option<Token> {
        self.tokens.last().map(|(t, _, _)| *t)
    }

    fn peek_next_non_ws(&self) -> Option<Token> {
        let mut rev_iter = self.tokens.iter().rev();
        rev_iter.next(); // Skip the first one.
        rev_iter
            .skip_while(|(t, _, _)| *t == Token::Whitespace || *t == Token::LineTerminator)
            .next()
            .map(|(t, _, _)| *t)
    }

    fn peek_next_str_non_ws(&self) -> Option<&'source str> {
        let mut rev_iter = self.tokens.iter().rev();
        rev_iter.next(); // Skip the first one.
        rev_iter
            .skip_while(|(t, _, _)| *t == Token::Whitespace || *t == Token::LineTerminator)
            .next()
            .map(|(_, s, _)| *s)
    }

    fn current_str(&self) -> Option<&'source str> {
        self.tokens.last().map(|(_, s, _)| *s)
    }

    fn current_pair(&self) -> Option<(Token, &'source str)> {
        self.tokens.last().map(|(t, s, _)| (*t, *s))
    }

    fn current_span(&self) -> Option<Span> {
        self.tokens.last().map(|(_, _, span)| span.clone())
    }

    fn error(&mut self, err: impl Into<String>) {
        self.errors.push((err.into(), self.current_span()));
    }

    fn bump_as(&mut self, kind: SyntaxKind) {
        let (_, text, _) = self.tokens.pop().unwrap();
        self.builder.token(kind.into(), text)
    }

    fn bump(&mut self) {
        let (kind, text, _) = self.tokens.pop().unwrap();
        self.builder.token(SyntaxKind::from(kind).into(), text);
    }

    fn skip_ws(&mut self) {
        // TODO: This probably needs to know how to deal with comments as well
        while self.current() == Some(Token::Whitespace)
            || self.current() == Some(Token::LineTerminator)
            || self.current() == Some(Token::Comma)
        {
            self.bump();
        }
    }
}

enum Res {
    Eof,
    Ok,
    ExpectedExecutableDef,
}

fn executable_def(parser: &mut Parser) -> Res {
    parser.skip_ws();
    match parser.current_pair() {
        None => {
            // Return some EOF indicator
            return Res::Eof;
        }
        Some((Token::Name, "query")) => {
            parser.builder.start_node(OPERATION_DEF.into());
            parser.bump_as(QUERY_KEYWORD);
            operation(parser);
            parser.builder.finish_node();
            return Res::Ok;
        }
        Some((Token::Name, "mutation")) => {
            parser.builder.start_node(OPERATION_DEF.into());
            parser.bump_as(MUTATION_KEYWORD);
            operation(parser);
            parser.builder.finish_node();
            return Res::Ok;
        }
        Some((Token::Name, "subscription")) => {
            parser.builder.start_node(OPERATION_DEF.into());
            parser.bump_as(SUBSCRIPTION_KEYWORD);
            operation(parser);
            parser.builder.finish_node();
            return Res::Ok;
        }
        Some((Token::OpenCurly, _)) => {
            parser.builder.start_node(OPERATION_DEF.into());
            selection_set(parser);
            parser.builder.finish_node();
            return Res::Ok;
        }
        Some((Token::Name, "fragment")) => {
            parser.builder.start_node(FRAGMENT_DEF.into());
            parser.bump_as(FRAGMENT_KEYWORD);
            fragment(parser);
            parser.builder.finish_node();
            return Res::Ok;
        }
        other => {
            return Res::ExpectedExecutableDef;
        }
    }
}

fn operation(parser: &mut Parser) {
    parser.skip_ws();
    if let Some(Token::Name) = parser.current() {
        parser.bump();
    }

    parser.skip_ws();
    if let Some(Token::OpenParen) = parser.current() {
        variable_defs(parser);
    }

    // TODO: directives

    parser.skip_ws();
    if let Some(Token::OpenCurly) = parser.current() {
        selection_set(parser);
    } else {
        todo!()
        // TODO: Error
    }
}

fn variable_defs(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::OpenParen));
    parser.builder.start_node(VARIABLE_DEFS.into());
    parser.bump();
    loop {
        parser.skip_ws();
        match parser.current() {
            None => break,
            Some(Token::Dollar) => {
                parser.builder.start_node(VARIABLE_DEF.into());
                variable(parser);
                parser.skip_ws();
                match parser.current() {
                    Some(Token::Colon) => {
                        parser.bump();
                    }
                    _ => parser.error("expected :"),
                }
                type_(parser);
                parser.skip_ws();
                if let Some(Token::Equals) = parser.current() {
                    parser.builder.start_node(DEFAULT_VALUE.into());
                    parser.bump();
                    constant_value(parser);
                    parser.builder.finish_node();
                }
                parser.builder.finish_node();
                // todo: default values
            }
            Some(Token::CloseParen) => {
                parser.bump();
                break;
            }
            _ => {
                todo!()
                // TODO: error
            }
        }
    }
    parser.builder.finish_node();
}

fn variable(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::Dollar));

    parser.builder.start_node(VARIABLE.into());
    parser.bump();
    match parser.current() {
        Some(Token::Name) => parser.bump(),
        _ => parser.error("expected name"),
    }
    parser.builder.finish_node();
}

fn type_(parser: &mut Parser) {
    parser.skip_ws();
    match parser.current() {
        Some(Token::Name) => {
            parser.builder.start_node(TYPE.into());
            named_type(parser);
            if let Some(Token::Bang) = parser.current() {
                parser.bump();
            }
            parser.builder.finish_node();
        }
        Some(Token::OpenSquare) => {
            parser.builder.start_node(TYPE.into());
            parser.bump();
            type_(parser);
            match parser.current() {
                Some(Token::CloseSquare) => {
                    parser.bump();
                }
                _ => {
                    parser.error("expected ]");
                }
            }
            parser.skip_ws();
            if let Some(Token::Bang) = parser.current() {
                parser.bump()
            }
            parser.builder.finish_node();
        }
        _ => {
            parser.error("expected name or [");
        }
    }
    parser.skip_ws();
}

fn named_type(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::Name));
    parser.builder.start_node(NAMED_TYPE.into());
    parser.bump();
    parser.builder.finish_node();
}

fn selection_set(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::OpenCurly));

    parser.builder.start_node(SELECTION_SET.into());
    parser.bump();

    loop {
        parser.skip_ws();
        match parser.current() {
            None => {
                parser.error("expected selection");
                break;
            }
            Some(Token::Name) => {
                parser.builder.start_node(SELECTION.into());
                field_selection(parser);
                parser.builder.finish_node();
            }
            Some(Token::CloseCurly) => {
                parser.bump();
                break;
            }
            Some(Token::Spread) => {
                parser.builder.start_node(SELECTION.into());
                match parser.peek_next_str_non_ws() {
                    Some("on") => inline_fragment(parser),
                    _ => fragment_spread(parser),
                }
                parser.builder.finish_node();
            }
            _ => {
                // TODO: is this good? not sure it is..
                parser.error("expected selection");
                break;
            }
        }
    }

    parser.builder.finish_node();
}

fn field_selection(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::Name));
    parser.builder.start_node(FIELD_SELECTION.into());
    if let Some(Token::Colon) = parser.peek_next_non_ws() {
        parser.builder.start_node(ALIAS.into());

        // Take the name & colon
        parser.bump();
        parser.skip_ws();
        parser.bump();
        parser.builder.finish_node();
    }

    parser.skip_ws();
    match parser.current() {
        Some(Token::Name) => {
            parser.bump();
        }
        _ => parser.error("expected name"),
    }

    parser.skip_ws();
    if let Some(Token::OpenParen) = parser.current() {
        arguments(parser);
    }

    parser.skip_ws();
    if let Some(Token::At) = parser.current() {
        todo!()
        // TODO: parse directives
    }

    parser.skip_ws();
    if let Some(Token::OpenCurly) = parser.current() {
        selection_set(parser)
    }

    parser.builder.finish_node();
}

#[cfg(test)]
mod tests {
    use std::{fs::File, io::Read};

    use rstest::rstest;

    use super::*;

    #[test]
    fn test_simple_parse() {
        let query = r#"
query {
    posts {
        title
        content
    }
}
"#;
        let result = parse(query);
        assert_eq!(result.errors, vec![]);
        assert_eq!(result.green_node.to_string(), query);
        insta::assert_debug_snapshot!(result.syntax());
    }

    #[rstest]
    #[case::fragment("tests/queries/fragment.graphql")]
    #[case::fragment_spread("tests/queries/fragment_spread.graphql")]
    #[case::inline_fragment("tests/queries/inline_fragment.graphql")]
    #[case::minimal("tests/queries/minimal.graphql")]
    #[case::minimal_mutation("tests/queries/minimal_mutation.graphql")]
    #[case::minimal_query("tests/queries/minimal_query.graphql")]
    #[case::named_query("tests/queries/named_query.graphql")]
    #[case::nested_selection("tests/queries/nested_selection.graphql")]
    #[case::query_aliases("tests/queries/query_aliases.graphql")]
    #[case::query_vars("tests/queries/query_vars.graphql")]
    #[case::query_arguments("tests/queries/query_arguments.graphql")]
    #[case::query_float_arguments("tests/queries/query_float_arguments.graphql")]
    #[case::query_list_argument("tests/queries/query_list_argument.graphql")]
    #[case::query_nameless_vars("tests/queries/query_nameless_vars.graphql")]
    // Commenting this one out because it requires comma ignores & comments
    //#[case::query_vars("tests/queries/query_nameless_vars_multiple_fields.graphql")]
    #[case::query_nameless_vars_multiple_fields_canonical(
        "tests/queries/query_nameless_vars_multiple_fields_canonical.graphql"
    )]
    #[case::query_object_argument("tests/queries/query_object_argument.graphql")]
    #[case::string_literal("tests/queries/string_literal.graphql")]
    #[case::triple_quoted_literal("tests/queries/triple_quoted_literal.graphql")]
    #[case::query_var_default_float("tests/queries/query_var_default_float.graphql")]
    #[case::query_var_default_list("tests/queries/query_var_default_list.graphql")]
    #[case::query_var_default_object("tests/queries/query_var_default_object.graphql")]
    #[case::query_var_default_string("tests/queries/query_var_default_string.graphql")]
    #[case::query_var_defaults("tests/queries/query_var_defaults.graphql")]
    fn test_query_file(#[case] file: String) {
        let mut query = String::new();
        File::open(file)
            .unwrap()
            .read_to_string(&mut query)
            .unwrap();

        let result = parse(&query);

        assert_eq!(result.errors, vec![]);
        assert_eq!(result.green_node.to_string(), query);
        // TODO: Determine the snapshot name from the file - to avoid ordering issues...
        insta::assert_debug_snapshot!(result.syntax());
    }
}
