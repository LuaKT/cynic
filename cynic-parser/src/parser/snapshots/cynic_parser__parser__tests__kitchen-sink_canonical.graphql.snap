---
source: cynic-parser/src/parser/mod.rs
expression: result.syntax()

---
DOCUMENT@0..841
  OPERATION_DEF@0..353
    OPERATION_TYPE@0..5
      QUERY_KEYWORD@0..5 "query"
    WHITESPACE@5..6 " "
    NAME@6..15 "queryName"
    VARIABLE_DEFS@15..56
      OPEN_PAREN@15..16 "("
      VARIABLE_DEF@16..35
        VARIABLE@16..20
          DOLLAR@16..17 "$"
          NAME@17..20 "foo"
        COLON@20..21 ":"
        WHITESPACE@21..22 " "
        TYPE@22..33
          NAMED_TYPE@22..33
            NAME@22..33 "ComplexType"
        COMMA@33..34 ","
        WHITESPACE@34..35 " "
      VARIABLE_DEF@35..55
        VARIABLE@35..40
          DOLLAR@35..36 "$"
          NAME@36..40 "site"
        COLON@40..41 ":"
        WHITESPACE@41..42 " "
        TYPE@42..46
          NAMED_TYPE@42..46
            NAME@42..46 "Site"
        WHITESPACE@46..47 " "
        DEFAULT_VALUE@47..55
          EQUALS@47..48 "="
          VALUE@48..55
            WHITESPACE@48..49 " "
            ENUM_VALUE@49..55
              NAME@49..55 "MOBILE"
      CLOSE_PAREN@55..56 ")"
    WHITESPACE@56..57 " "
    SELECTION_SET@57..353
      OPEN_CURLY@57..58 "{"
      LINE_TERMINATOR@58..59 "\n"
      WHITESPACE@59..61 "  "
      FIELD_SELECTION@61..351
        ALIAS@61..74
          NAME@61..73 "whoever123is"
          COLON@73..74 ":"
        WHITESPACE@74..75 " "
        NAME@75..79 "node"
        ARGUMENTS@79..96
          OPEN_PAREN@79..80 "("
          ARGUMENT@80..94
            NAME@80..82 "id"
            COLON@82..83 ":"
            VALUE@83..94
              WHITESPACE@83..84 " "
              LIST_VALUE@84..94
                OPEN_SQUARE@84..85 "["
                VALUE@85..90
                  INTEGER_VALUE@85..88
                    NUMBER@85..88 "123"
                  COMMA@88..89 ","
                  WHITESPACE@89..90 " "
                VALUE@90..93
                  INTEGER_VALUE@90..93
                    NUMBER@90..93 "456"
                CLOSE_SQUARE@93..94 "]"
          CLOSE_PAREN@94..95 ")"
          WHITESPACE@95..96 " "
        SELECTION_SET@96..351
          OPEN_CURLY@96..97 "{"
          LINE_TERMINATOR@97..98 "\n"
          WHITESPACE@98..102 "    "
          FIELD_SELECTION@102..109
            NAME@102..104 "id"
            LINE_TERMINATOR@104..105 "\n"
            WHITESPACE@105..109 "    "
          INLINE_FRAGMENT@109..277
            SPREAD@109..112 "..."
            WHITESPACE@112..113 " "
            TYPE_CONDITION@113..120
              ON@113..115 "on"
              WHITESPACE@115..116 " "
              NAMED_TYPE@116..120
                NAME@116..120 "User"
            WHITESPACE@120..121 " "
            DIRECTIVES@121..128
              DIRECTIVE@121..128
                AT@121..122 "@"
                NAME@122..127 "defer"
                WHITESPACE@127..128 " "
            SELECTION_SET@128..277
              OPEN_CURLY@128..129 "{"
              LINE_TERMINATOR@129..130 "\n"
              WHITESPACE@130..136 "      "
              FIELD_SELECTION@136..271
                NAME@136..142 "field2"
                WHITESPACE@142..143 " "
                SELECTION_SET@143..271
                  OPEN_CURLY@143..144 "{"
                  LINE_TERMINATOR@144..145 "\n"
                  WHITESPACE@145..153 "        "
                  FIELD_SELECTION@153..164
                    NAME@153..155 "id"
                    LINE_TERMINATOR@155..156 "\n"
                    WHITESPACE@156..164 "        "
                  FIELD_SELECTION@164..263
                    ALIAS@164..170
                      NAME@164..169 "alias"
                      COLON@169..170 ":"
                    WHITESPACE@170..171 " "
                    NAME@171..177 "field1"
                    ARGUMENTS@177..202
                      OPEN_PAREN@177..178 "("
                      ARGUMENT@178..189
                        NAME@178..183 "first"
                        COLON@183..184 ":"
                        VALUE@184..189
                          WHITESPACE@184..185 " "
                          INTEGER_VALUE@185..187
                            NUMBER@185..187 "10"
                          COMMA@187..188 ","
                          WHITESPACE@188..189 " "
                      ARGUMENT@189..200
                        NAME@189..194 "after"
                        COLON@194..195 ":"
                        VALUE@195..200
                          WHITESPACE@195..196 " "
                          VARIABLE@196..200
                            DOLLAR@196..197 "$"
                            NAME@197..200 "foo"
                      CLOSE_PAREN@200..201 ")"
                      WHITESPACE@201..202 " "
                    DIRECTIVES@202..221
                      DIRECTIVE@202..221
                        AT@202..203 "@"
                        NAME@203..210 "include"
                        ARGUMENTS@210..221
                          OPEN_PAREN@210..211 "("
                          ARGUMENT@211..219
                            NAME@211..213 "if"
                            COLON@213..214 ":"
                            VALUE@214..219
                              WHITESPACE@214..215 " "
                              VARIABLE@215..219
                                DOLLAR@215..216 "$"
                                NAME@216..219 "foo"
                          CLOSE_PAREN@219..220 ")"
                          WHITESPACE@220..221 " "
                    SELECTION_SET@221..263
                      OPEN_CURLY@221..222 "{"
                      LINE_TERMINATOR@222..223 "\n"
                      WHITESPACE@223..233 "          "
                      FIELD_SELECTION@233..246
                        NAME@233..235 "id"
                        LINE_TERMINATOR@235..236 "\n"
                        WHITESPACE@236..246 "          "
                      FRAGMENT_SPREAD@246..262
                        SPREAD@246..249 "..."
                        FRAGMENT_NAME@249..253
                          NAME@249..253 "frag"
                        LINE_TERMINATOR@253..254 "\n"
                        WHITESPACE@254..262 "        "
                      CLOSE_CURLY@262..263 "}"
                  LINE_TERMINATOR@263..264 "\n"
                  WHITESPACE@264..270 "      "
                  CLOSE_CURLY@270..271 "}"
              LINE_TERMINATOR@271..272 "\n"
              WHITESPACE@272..276 "    "
              CLOSE_CURLY@276..277 "}"
          LINE_TERMINATOR@277..278 "\n"
          WHITESPACE@278..282 "    "
          FRAGMENT_SPREAD@282..322
            SPREAD@282..285 "..."
            WHITESPACE@285..286 " "
            DIRECTIVES@286..306
              DIRECTIVE@286..306
                AT@286..287 "@"
                NAME@287..291 "skip"
                ARGUMENTS@291..306
                  OPEN_PAREN@291..292 "("
                  ARGUMENT@292..304
                    NAME@292..298 "unless"
                    COLON@298..299 ":"
                    VALUE@299..304
                      WHITESPACE@299..300 " "
                      VARIABLE@300..304
                        DOLLAR@300..301 "$"
                        NAME@301..304 "foo"
                  CLOSE_PAREN@304..305 ")"
                  WHITESPACE@305..306 " "
            SELECTION_SET@306..322
              OPEN_CURLY@306..307 "{"
              LINE_TERMINATOR@307..308 "\n"
              WHITESPACE@308..314 "      "
              FIELD_SELECTION@314..321
                NAME@314..316 "id"
                LINE_TERMINATOR@316..317 "\n"
                WHITESPACE@317..321 "    "
              CLOSE_CURLY@321..322 "}"
          LINE_TERMINATOR@322..323 "\n"
          WHITESPACE@323..327 "    "
          FRAGMENT_SPREAD@327..347
            SPREAD@327..330 "..."
            WHITESPACE@330..331 " "
            SELECTION_SET@331..347
              OPEN_CURLY@331..332 "{"
              LINE_TERMINATOR@332..333 "\n"
              WHITESPACE@333..339 "      "
              FIELD_SELECTION@339..346
                NAME@339..341 "id"
                LINE_TERMINATOR@341..342 "\n"
                WHITESPACE@342..346 "    "
              CLOSE_CURLY@346..347 "}"
          LINE_TERMINATOR@347..348 "\n"
          WHITESPACE@348..350 "  "
          CLOSE_CURLY@350..351 "}"
      LINE_TERMINATOR@351..352 "\n"
      CLOSE_CURLY@352..353 "}"
  LINE_TERMINATOR@353..354 "\n"
  LINE_TERMINATOR@354..355 "\n"
  OPERATION_DEF@355..436
    OPERATION_TYPE@355..363
      MUTATION_KEYWORD@355..363 "mutation"
    WHITESPACE@363..364 " "
    NAME@364..373 "likeStory"
    WHITESPACE@373..374 " "
    SELECTION_SET@374..436
      OPEN_CURLY@374..375 "{"
      LINE_TERMINATOR@375..376 "\n"
      WHITESPACE@376..378 "  "
      FIELD_SELECTION@378..434
        NAME@378..382 "like"
        ARGUMENTS@382..395
          OPEN_PAREN@382..383 "("
          ARGUMENT@383..393
            NAME@383..388 "story"
            COLON@388..389 ":"
            VALUE@389..393
              WHITESPACE@389..390 " "
              INTEGER_VALUE@390..393
                NUMBER@390..393 "123"
          CLOSE_PAREN@393..394 ")"
          WHITESPACE@394..395 " "
        DIRECTIVES@395..402
          DIRECTIVE@395..402
            AT@395..396 "@"
            NAME@396..401 "defer"
            WHITESPACE@401..402 " "
        SELECTION_SET@402..434
          OPEN_CURLY@402..403 "{"
          LINE_TERMINATOR@403..404 "\n"
          WHITESPACE@404..408 "    "
          FIELD_SELECTION@408..430
            NAME@408..413 "story"
            WHITESPACE@413..414 " "
            SELECTION_SET@414..430
              OPEN_CURLY@414..415 "{"
              LINE_TERMINATOR@415..416 "\n"
              WHITESPACE@416..422 "      "
              FIELD_SELECTION@422..429
                NAME@422..424 "id"
                LINE_TERMINATOR@424..425 "\n"
                WHITESPACE@425..429 "    "
              CLOSE_CURLY@429..430 "}"
          LINE_TERMINATOR@430..431 "\n"
          WHITESPACE@431..433 "  "
          CLOSE_CURLY@433..434 "}"
      LINE_TERMINATOR@434..435 "\n"
      CLOSE_CURLY@435..436 "}"
  LINE_TERMINATOR@436..437 "\n"
  LINE_TERMINATOR@437..438 "\n"
  OPERATION_DEF@438..648
    OPERATION_TYPE@438..450
      SUBSCRIPTION_KEYWORD@438..450 "subscription"
    WHITESPACE@450..451 " "
    NAME@451..472 "StoryLikeSubscription"
    VARIABLE_DEFS@472..505
      OPEN_PAREN@472..473 "("
      VARIABLE_DEF@473..504
        VARIABLE@473..479
          DOLLAR@473..474 "$"
          NAME@474..479 "input"
        COLON@479..480 ":"
        WHITESPACE@480..481 " "
        TYPE@481..504
          NAMED_TYPE@481..504
            NAME@481..504 "StoryLikeSubscribeInput"
      CLOSE_PAREN@504..505 ")"
    WHITESPACE@505..506 " "
    SELECTION_SET@506..648
      OPEN_CURLY@506..507 "{"
      LINE_TERMINATOR@507..508 "\n"
      WHITESPACE@508..510 "  "
      FIELD_SELECTION@510..646
        NAME@510..528 "storyLikeSubscribe"
        ARGUMENTS@528..544
          OPEN_PAREN@528..529 "("
          ARGUMENT@529..542
            NAME@529..534 "input"
            COLON@534..535 ":"
            VALUE@535..542
              WHITESPACE@535..536 " "
              VARIABLE@536..542
                DOLLAR@536..537 "$"
                NAME@537..542 "input"
          CLOSE_PAREN@542..543 ")"
          WHITESPACE@543..544 " "
        SELECTION_SET@544..646
          OPEN_CURLY@544..545 "{"
          LINE_TERMINATOR@545..546 "\n"
          WHITESPACE@546..550 "    "
          FIELD_SELECTION@550..642
            NAME@550..555 "story"
            WHITESPACE@555..556 " "
            SELECTION_SET@556..642
              OPEN_CURLY@556..557 "{"
              LINE_TERMINATOR@557..558 "\n"
              WHITESPACE@558..564 "      "
              FIELD_SELECTION@564..594
                NAME@564..570 "likers"
                WHITESPACE@570..571 " "
                SELECTION_SET@571..594
                  OPEN_CURLY@571..572 "{"
                  LINE_TERMINATOR@572..573 "\n"
                  WHITESPACE@573..581 "        "
                  FIELD_SELECTION@581..593
                    NAME@581..586 "count"
                    LINE_TERMINATOR@586..587 "\n"
                    WHITESPACE@587..593 "      "
                  CLOSE_CURLY@593..594 "}"
              LINE_TERMINATOR@594..595 "\n"
              WHITESPACE@595..601 "      "
              FIELD_SELECTION@601..636
                NAME@601..613 "likeSentence"
                WHITESPACE@613..614 " "
                SELECTION_SET@614..636
                  OPEN_CURLY@614..615 "{"
                  LINE_TERMINATOR@615..616 "\n"
                  WHITESPACE@616..624 "        "
                  FIELD_SELECTION@624..635
                    NAME@624..628 "text"
                    LINE_TERMINATOR@628..629 "\n"
                    WHITESPACE@629..635 "      "
                  CLOSE_CURLY@635..636 "}"
              LINE_TERMINATOR@636..637 "\n"
              WHITESPACE@637..641 "    "
              CLOSE_CURLY@641..642 "}"
          LINE_TERMINATOR@642..643 "\n"
          WHITESPACE@643..645 "  "
          CLOSE_CURLY@645..646 "}"
      LINE_TERMINATOR@646..647 "\n"
      CLOSE_CURLY@647..648 "}"
  LINE_TERMINATOR@648..649 "\n"
  LINE_TERMINATOR@649..650 "\n"
  FRAGMENT_DEF@650..773
    FRAGMENT_KEYWORD@650..658 "fragment"
    WHITESPACE@658..659 " "
    FRAGMENT_NAME@659..663
      NAME@659..663 "frag"
    WHITESPACE@663..664 " "
    TYPE_CONDITION@664..673
      ON@664..666 "on"
      WHITESPACE@666..667 " "
      NAMED_TYPE@667..673
        NAME@667..673 "Friend"
    WHITESPACE@673..674 " "
    SELECTION_SET@674..773
      OPEN_CURLY@674..675 "{"
      LINE_TERMINATOR@675..676 "\n"
      WHITESPACE@676..678 "  "
      FIELD_SELECTION@678..772
        NAME@678..681 "foo"
        ARGUMENTS@681..772
          OPEN_PAREN@681..682 "("
          ARGUMENT@682..695
            NAME@682..686 "size"
            COLON@686..687 ":"
            VALUE@687..695
              WHITESPACE@687..688 " "
              VARIABLE@688..693
                DOLLAR@688..689 "$"
                NAME@689..693 "size"
              COMMA@693..694 ","
              WHITESPACE@694..695 " "
          ARGUMENT@695..704
            NAME@695..698 "bar"
            COLON@698..699 ":"
            VALUE@699..704
              WHITESPACE@699..700 " "
              VARIABLE@700..702
                DOLLAR@700..701 "$"
                NAME@701..702 "b"
              COMMA@702..703 ","
              WHITESPACE@703..704 " "
          ARGUMENT@704..770
            NAME@704..707 "obj"
            COLON@707..708 ":"
            VALUE@708..770
              WHITESPACE@708..709 " "
              OBJECT_VALUE@709..770
                OPEN_CURLY@709..710 "{"
                OBJECT_FIELD@710..757
                  NAME@710..715 "block"
                  COLON@715..716 ":"
                  VALUE@716..757
                    WHITESPACE@716..717 " "
                    STRING_VALUE@717..755
                      BLOCK_QUOTE@717..720 "\"\"\""
                      STRING_CONTENTS@720..721 "\n"
                      STRING_CONTENTS@721..722 "\n"
                      STRING_CONTENTS@722..726 "    "
                      STRING_CONTENTS@726..731 "block"
                      STRING_CONTENTS@731..732 " "
                      STRING_CONTENTS@732..738 "string"
                      STRING_CONTENTS@738..739 " "
                      STRING_CONTENTS@739..743 "uses"
                      STRING_CONTENTS@743..744 " "
                      STRING_CONTENTS@744..746 "\\\""
                      STRING_CONTENTS@746..747 "\""
                      STRING_CONTENTS@747..748 "\""
                      STRING_CONTENTS@748..749 "\n"
                      STRING_CONTENTS@749..750 "\n"
                      STRING_CONTENTS@750..752 "  "
                      BLOCK_QUOTE@752..755 "\"\"\""
                    COMMA@755..756 ","
                    WHITESPACE@756..757 " "
                OBJECT_FIELD@757..769
                  NAME@757..760 "key"
                  COLON@760..761 ":"
                  VALUE@761..769
                    WHITESPACE@761..762 " "
                    STRING_VALUE@762..769
                      QUOTE@762..763 "\""
                      STRING_CONTENTS@763..768 "value"
                      QUOTE@768..769 "\""
                CLOSE_CURLY@769..770 "}"
          CLOSE_PAREN@770..771 ")"
          LINE_TERMINATOR@771..772 "\n"
      CLOSE_CURLY@772..773 "}"
  LINE_TERMINATOR@773..774 "\n"
  LINE_TERMINATOR@774..775 "\n"
  OPERATION_DEF@775..840
    SELECTION_SET@775..840
      OPEN_CURLY@775..776 "{"
      LINE_TERMINATOR@776..777 "\n"
      WHITESPACE@777..779 "  "
      FIELD_SELECTION@779..833
        NAME@779..786 "unnamed"
        ARGUMENTS@786..833
          OPEN_PAREN@786..787 "("
          ARGUMENT@787..801
            NAME@787..793 "truthy"
            COLON@793..794 ":"
            VALUE@794..801
              WHITESPACE@794..795 " "
              BOOL_VALUE@795..799
                TRUE@795..799 "true"
              COMMA@799..800 ","
              WHITESPACE@800..801 " "
          ARGUMENT@801..816
            NAME@801..807 "falsey"
            COLON@807..808 ":"
            VALUE@808..816
              WHITESPACE@808..809 " "
              BOOL_VALUE@809..814
                FALSE@809..814 "false"
              COMMA@814..815 ","
              WHITESPACE@815..816 " "
          ARGUMENT@816..829
            NAME@816..823 "nullish"
            COLON@823..824 ":"
            VALUE@824..829
              WHITESPACE@824..825 " "
              NULL@825..829 "null"
          CLOSE_PAREN@829..830 ")"
          LINE_TERMINATOR@830..831 "\n"
          WHITESPACE@831..833 "  "
      FIELD_SELECTION@833..839
        NAME@833..838 "query"
        LINE_TERMINATOR@838..839 "\n"
      CLOSE_CURLY@839..840 "}"
  LINE_TERMINATOR@840..841 "\n"

